modules/gapi/src/streaming/onevpl/cfg_param_device_selector.cpp

```diff
@@ -0,0 +1,296 @@
+            {
+                ID3D11Multithread       *pD11Multithread;
+                device_context->QueryInterface(IID_PPV_ARGS(&pD11Multithread));
+                pD11Multithread->SetMultithreadProtected(true);
+                pD11Multithread->Release();
+            }
+
+            suggested_device = IDeviceSelector::create<Device>(hw_handle, AccelType::DX11);
+            suggested_context = IDeviceSelector::create<Context>(device_context, AccelType::DX11);
+#else
+            GAPI_LOG_WARNING(nullptr, "Unavailable \"mfxImplDescription.AccelerationMode: MFX_ACCEL_MODE_VIA_D3D11\""
+                                      "was choosed for current project configuration");
+            throw std::logic_error("Unsupported \"mfxImplDescription.AccelerationMode: MFX_ACCEL_MODE_VIA_D3D11\"");
+#endif // HAVE_DIRECTX
+#endif // HAVE_D3D11
+            break;
+        }
+        case MFX_ACCEL_MODE_NA: {
+            // nothing to do
+            break;
+        }
+        default:
+            throw std::logic_error("Unsupported \"mfxImplDescription.AccelerationMode\" requested: " +
+                                   std::to_string(accel_mode.Data.U32));
+            break;
+    }
+}
+
+CfgParamDeviceSelector::CfgParamDeviceSelector(Device::Ptr device_ptr,
+                                               Context::Ptr ctx_ptr,
+                                               const CfgParams& cfg_params) :
+    IDeviceSelector(),
+    suggested_device(IDeviceSelector::create<Device>(nullptr, AccelType::HOST)),
+    suggested_context(IDeviceSelector::create<Context>(nullptr, AccelType::HOST)) {
+    auto accel_mode_it =
+        std::find_if(cfg_params.begin(), cfg_params.end(), [] (const CfgParam& value) {
+            return value.get_name() == "mfxImplDescription.AccelerationMode";
+        });
+    if (accel_mode_it == cfg_params.end()) {
+        GAPI_LOG_WARNING(nullptr, "Cannot deternime \"device_ptr\" type. "
+                         "Make sure a param \"mfxImplDescription.AccelerationMode\" "
+                         "presents in configurations and has correct value according to "
+                         "\"device_ptr\" type");
+        throw std::logic_error("Missing \"mfxImplDescription.AccelerationMode\" param");
+    }
+
+    GAPI_LOG_DEBUG(nullptr, "Turn on HW acceleration support for device: " <<
+                            device_ptr <<
+                            ", context: " << ctx_ptr);
+    if (!device_ptr) {
+        GAPI_LOG_WARNING(nullptr, "Empty \"device_ptr\" is not allowed when "
+                         "param \"mfxImplDescription.AccelerationMode\" existed");
+        throw std::logic_error("Invalid param: \"device_ptr\"");
+    }
+
+     if (!ctx_ptr) {
+        GAPI_LOG_WARNING(nullptr, "Empty \"ctx_ptr\" is not allowed");
+        throw std::logic_error("Invalid  param: \"ctx_ptr\"");
+    }
+    mfxVariant accel_mode = cfg_param_to_mfx_variant(*accel_mode_it);
+
+    switch(accel_mode.Data.U32) {
```

Comment:
this case is handled in previous constuctor, too - should that logic be unified?

Reply:
the first ctor uses device creation from default adapter - no external pointers from user
but this ctor just wrap user-created device passed by `void*` pointer AND uses actual device type through `CfgParams` to restore type and work out some checks